File: app/src/main/java/com/hypertrack/android/ui/screens/sign_in/SignInFragment.java
package com.hypertrack.android.ui.screens.sign_in;

import android.os.Bundle;
import android.text.Editable;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.EditText;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.navigation.fragment.NavHostFragment;

import com.amazonaws.services.cognitoidentityprovider.model.NotAuthorizedException;
import com.amazonaws.services.cognitoidentityprovider.model.UserNotConfirmedException;
import com.amazonaws.services.cognitoidentityprovider.model.UserNotFoundException;
import com.hypertrack.android.cognito.CognitoClient;
import com.hypertrack.android.live.App;
import com.hypertrack.android.live.HTTextWatcher;
import com.hypertrack.android.ui.base.BaseFragment;
import com.hypertrack.android.ui.base.ProgressDialogFragment;
import com.hypertrack.android.ui.screens.sign_up.SignUpFragment;
import com.hypertrack.logistics.android.github.R;

public class SignInFragment extends ProgressDialogFragment implements CognitoClient.Callback {

    private static final String TAG = App.TAG + "SignInFragment";

    private EditText emailAddressEditText;
    private EditText passwordEditText;
    private View passwordClear;
    private TextView incorrect;

    public SignInFragment() {
        super(R.layout.fragment_signin);
    }


    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        emailAddressEditText = view.findViewById(R.id.email_address);
        passwordEditText = view.findViewById(R.id.password);
        passwordClear = view.findViewById(R.id.password_clear);
        incorrect = view.findViewById(R.id.incorrect);

        //todo task
//        emailAddressEditText.setText("ozh14873@cuoly.com");
//        passwordEditText.setText("qwerty123");

        passwordEditText.addTextChangedListener(new HTTextWatcher() {
            @Override
            public void afterTextChanged(Editable editable) {
                passwordClear.setVisibility(TextUtils.isEmpty(editable) ? View.INVISIBLE : View.VISIBLE);
            }
        });
        passwordClear.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                passwordEditText.setText("");
            }
        });

        view.findViewById(R.id.sign_in).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                incorrect.setText("");

                String email = emailAddressEditText.getText().toString();
                String password = passwordEditText.getText().toString();
                if (!TextUtils.isEmpty(email) && !TextUtils.isEmpty(password)) {
                    startSignIn(email, password);
                }
            }
        });

        view.findViewById(R.id.sign_up).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                NavHostFragment.findNavController(SignInFragment.this)
                        .navigate(SignInFragmentDirections.Companion.actionSignInFragmentToSignUpFragment());
            }
        });
    }

    public void startSignIn(String email, String password) {
        showProgress();
        CognitoClient.getInstance(getContext()).signIn(email, password, SignInFragment.this);
    }

    @Override
    public void onSuccess(CognitoClient mobileClient) {
        if (getActivity() != null) {
            dismissProgress();
            //todo task repo
            NavHostFragment.findNavController(SignInFragment.this)
                    .navigate(SignInFragmentDirections.Companion.actionSignInFragmentToVisitManagementFragment());
//            ((LaunchActivity)getActivity()).onLoginCompleted();
        }
    }

    @Override
    public void onError(String message, Exception e) {
        if (getActivity() != null) {
            dismissProgress();

            if (e instanceof UserNotConfirmedException) {
                NavHostFragment.findNavController(SignInFragment.this)
                        .navigate(SignInFragmentDirections.Companion.actionSignInFragmentToConfirmFragment());
            } else if (e instanceof UserNotFoundException) {
                incorrect.setText(R.string.user_does_not_exist);
            } else if (e instanceof NotAuthorizedException) {
                incorrect.setText(R.string.incorrect_username_or_pass);
            } else {
                Toast.makeText(getActivity(), message, Toast.LENGTH_LONG).show();
            }
        }
    }
}

File: app/src/main/java/com/hypertrack/android/ui/screens/sign_in/SignInFragment.kt
package com.hypertrack.android.ui.screens.sign_in

import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.view.View
import androidx.fragment.app.viewModels
import androidx.navigation.fragment.findNavController
import androidx.navigation.fragment.navArgs
import com.hypertrack.android.ui.base.ProgressDialogFragment
import com.hypertrack.android.utils.MyApplication
import com.hypertrack.logistics.android.github.R
import kotlinx.android.synthetic.main.fragment_signin.*

class SignInFragment : ProgressDialogFragment(R.layout.fragment_signin) {

    private val args: SignInFragmentArgs by navArgs()

    private val vm: SignInViewModel by viewModels {
        MyApplication.injector.provideViewModelFactory(MyApplication.context)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        email_address.addTextChangedListener(object : TextWatcher {
            override fun afterTextChanged(s: Editable?) {}

            override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {}

            override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {
                s?.let { vm.onLoginTextChanged(it) }
            }
        })

        password.addTextChangedListener(object : TextWatcher {
            override fun afterTextChanged(s: Editable?) {}

            override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {}

            override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {
                s?.let { vm.onPasswordTextChanged(it) }
            }
        })

        args.email?.let {
            email_address.setText(it)
        }

        sign_up.setOnClickListener {
            vm.onSignUpClick()
        }

        vm.destination.observe(viewLifecycleOwner) {
            findNavController().navigate(it)
        }

        vm.errorText.observe(viewLifecycleOwner, {
            incorrect.text = it
        })


        sign_in.setOnClickListener { vm.onLoginClick(mainActivity()) }

        vm.showProgress.observe(viewLifecycleOwner) { show ->
            if (show) showProgress() else dismissProgress()
        }

        //todo rounded corners
        vm.isLoginButtonClickable.observe(viewLifecycleOwner) { isClickable ->
            // Log.d(TAG, "Setting login button clickability $isClickable")
            sign_in.isEnabled = isClickable
            sign_in.setBackgroundColor(
                if (isClickable)
                    requireContext().getColor(R.color.colorHyperTrackGreen)
                else
                    requireContext().getColor(R.color.colorBtnDisable)
            )
        }

    }

    companion object {
        const val TAG = "AccountLoginAct"
    }
}